### PR TITLE
[bug] - fix data races

### PR DESCRIPTION
**Pull request by ahrav**
- Add test to catch all the data races below
- Multiple repos can concurrently use the git source, therefore we need to make the metrics counter atomic.
NOTE: We should probably look for a less hacky solution for the long term. This is just to temporarily fix the issues.

Screenshot 2023-07-29 at 3 08 18 PM

- Modifying the user and token also needs to be protected
Screenshot 2023-07-29 at 7 16 45 PM

- Modifying the scanOptions also need protection
Screenshot 2023-07-29 at 7 29 19 PM

### CODE REVIEW

File: pkg/sources/github/github.go [hunk 1]
```diff
@@ -78,6 +86,13 @@ func (s *Source) WithScanOptions(scanOptions *git.ScanOptions) {
 	s.scanOptions = scanOptions
 }
 
+func (s *Source) setScanOptions(base, head string) {
+	s.scanOptMu.Lock()
+	defer s.scanOptMu.Unlock()
+	s.scanOptions.BaseHash = base
+	s.scanOptions.BaseHash = head
```
Comment (rosecodym) [inline pkg/sources/github/github.go:93]: i'm guessing there's a typo here

File: pkg/sources/github/github.go [hunk 2]
```diff
@@ -678,6 +693,7 @@ func (s *Source) scan(ctx context.Context, installationClient *github.Client, ch
 			path, repo, err = s.cloneRepo(ctx, repoURL, installationClient)
 			if err != nil {
 				scanErrs.Add(err)
+				return nil
```
Comment (rosecodym) [inline pkg/sources/github/github.go:696]: is this related to the data race?

File: pkg/sources/github/repo.go [hunk 3]
```diff
@@ -50,12 +50,15 @@ func (s *Source) cloneRepo(
 
 	case *sourcespb.GitHub_Token:
 		// We never refresh user provided tokens, so if we already have them, we never need to try and fetch them again.
+		s.userMu.Lock()
```
Comment (rosecodym) [inline pkg/sources/github/repo.go:53]: If `s.userAndToken` panics, this will be locked forever, right? Do you think it's worth breaking out a new function (inline or otherwise) so you can use `defer`?

**Review by rosecodym — CHANGES_REQUESTED** [review]
i think there's a small typo in addition to my other non-blocking questions

File: pkg/sources/github/github.go (hunk 2 above)
Comment (ahrav) [reply pkg/sources/github/github.go:696]: This was causing a test to fail. If we can't clone a repo we should not try to scan.

**Review by rosecodym — APPROVED** [review]
